### PR TITLE
util/coll: Create collective address for an av_set correctly

### PR DESCRIPTION
**Pull request by zhenggb72**
ofi_av_set_addr creates a new collective address for an av_set
by building a new coll_mc with only the group members, instead
of using the av_set's AV address.

Signed-off-by: Gengbin Zheng <gengbin.zheng@intel.com>

fixes https://github.com/ofiwg/libfabric/issues/5809

### CODE REVIEW

File: prov/util/src/util_coll.c [hunk 1]
```diff
@@ -764,6 +754,29 @@ static struct fi_ops util_coll_fi_ops = {
...
+	coll_mc->mc_fid.fid.context = NULL;
+	coll_mc->mc_fid.fid.ops = &util_coll_fi_ops;
+	coll_mc->mc_fid.fi_addr = (uintptr_t) coll_mc;
+	coll_mc->av_set = av_set;
+	coll_mc->ep = NULL;
+
+	*coll_addr = (uintptr_t)coll_mc;
+
+	return FI_SUCCESS;
+}
```
Comment (shefty) [inline prov/util/src/util_coll.c:778]: I have a couple requests while you're in this part of the code.  Can you either delete the util_coll_mc_alloc() routine, or move the above functionality there?  As it is now, it does nothing but wrap calloc.

Also, replace the duplicate allocation code (for new_coll_mc) in fi_join_multicast with whatever new function (even calling ofi_av_set_addr) allocates the structure.  Likewise, it looks like there's duplicate allocation code in util_coll_av_init().

Basically, there should only be 1 place where we allocate struct util_coll_mc.

Comment (zhenggb72) [reply prov/util/src/util_coll.c:783]: fixed, please review.
Also add another commit to add coll_mc to av_set struct.

File: include/ofi_coll.h [hunk 2]
```diff
@@ -67,6 +67,7 @@ struct util_av_set {
 	fi_addr_t		*fi_addr_array;
 	size_t			fi_addr_count;
 	uint64_t		flags;
+	struct util_coll_mc     *coll_mc;
```
Comment (shefty) [inline include/ofi_coll.h:70]: We can probably just embed the structure, rather than storing a pointer and needing to alloc/free it.

Doing that, util_coll_mc_alloc() should be renamed to util_coll_mc_init(), with the calloc pulled out.  The caller (only place would be fi_join_multicast) would allocate the structure, then call init() to initialize it.

File: prov/util/src/util_coll.c [hunk 3]
```diff
@@ -1003,17 +1021,17 @@ int ofi_join_collective(struct fid_ep *ep, fi_addr_t coll_addr,
...
 
-	ret = util_coll_mc_alloc(&new_coll_mc);
-	if (ret)
-		return ret;
+	new_coll_mc = av_set->coll_mc;
+	if (!new_coll_mc) {
+		ret = util_coll_mc_alloc(&new_coll_mc);
+		if (ret)
+			goto err1;
+	}
```
Comment (shefty) [inline prov/util/src/util_coll.c:1029]: We always need to allocate a new util_coll_mc here.  This structure is returned to the caller through the mc parameter and must be closed by the app via fi_close().  Along this line, when we allocate the util_coll_mc, we should increment a reference count on the av_set.  Closing the mc should decrement the reference count.  I see that av_set has a ref count available, but I don't see that it's getting incremented by this function.  The decrement should be in util_coll_close().  Since the util_coll_mc also references the endpoint, references should be acquired and released on the endpoint as well.

You mentioned this, but the av_set->av_set_fid.fid.ops are missing.  That's where an av_set_close() function would be set.  Close should check that the ref count is 0, and fail otherwise, before freeing the av_set.  To set this, look at util_coll_fi_ops as an example.  You just need to define a util_av_set_close() function that checks that the ref count is 0, and frees the av_set.  Then setup a static struct (similar to util_coll_fi_ops) that sets the function pointers.  Finally, assign av_set->av_set_fid.fid_ops = the_name_of_the_static_struct inside the av_set creation call below.

File: prov/util/src/util_coll.c [hunk 4]
```diff
@@ -764,6 +745,43 @@ static struct fi_ops util_coll_fi_ops = {
...
+
+	return FI_SUCCESS;
+}
+
+
+int ofi_av_set_addr(struct fid_av_set *set, fi_addr_t *coll_addr)
+{
+	struct util_av_set *av_set;
+	struct util_coll_mc *coll_mc;
+        int ret;
```
Comment (shefty) [inline prov/util/src/util_coll.c:767]: nit: spacing is off here

File: prov/util/src/util_coll.c [hunk 5]
```diff
@@ -764,6 +745,43 @@ static struct fi_ops util_coll_fi_ops = {
...
+
+	(*coll_mc)->mc_fid.fid.fclass = FI_CLASS_MC;
+	(*coll_mc)->mc_fid.fid.context = NULL;
+	(*coll_mc)->mc_fid.fid.ops = &util_coll_fi_ops;
+	(*coll_mc)->mc_fid.fi_addr = (uintptr_t)* coll_mc;
+
+	return FI_SUCCESS;
+}
+
+
```
Comment (shefty) [inline prov/util/src/util_coll.c:762]: nit: extra blank line

File: prov/util/src/util_coll.c [hunk 6]
```diff
@@ -764,6 +745,43 @@ static struct fi_ops util_coll_fi_ops = {
 	.ops_open = fi_no_ops_open,
 };
 
+static inline int util_coll_mc_alloc(struct util_coll_mc **coll_mc)
```
Comment (shefty) [inline prov/util/src/util_coll.c:748]: I would pass in the av_set and endpoint into this function, possibly the context as well.  See above comment about changing this from alloc() to init().

File: prov/util/src/util_coll.c [hunk 7]
```diff
@@ -764,6 +745,43 @@ static struct fi_ops util_coll_fi_ops = {
...
+{
+	*coll_mc = calloc(1, sizeof(**coll_mc));
+	if (!*coll_mc)
+		return -FI_ENOMEM;
+
+	(*coll_mc)->mc_fid.fid.fclass = FI_CLASS_MC;
+	(*coll_mc)->mc_fid.fid.context = NULL;
+	(*coll_mc)->mc_fid.fid.ops = &util_coll_fi_ops;
+	(*coll_mc)->mc_fid.fi_addr = (uintptr_t)* coll_mc;
+
```
Comment (shefty) [inline prov/util/src/util_coll.c:758]: Here we should save the av_set and ep, and bump reference counts on both.  The references should be dropped prior to freeing the util_coll_mc.

File: prov/util/src/util_coll.c [hunk 8]
```diff
@@ -764,6 +745,43 @@ static struct fi_ops util_coll_fi_ops = {
...
+	coll_mc = av_set->coll_mc;
+	if (!coll_mc) {
+		ret = util_coll_mc_alloc(&coll_mc);
+		if (ret)
+			return ret;
+	}
+
+	coll_mc->av_set = av_set;
+
+	*coll_addr = (uintptr_t)coll_mc;
```
Comment (shefty) [inline prov/util/src/util_coll.c:780]: This function should just need to do:
```
coll_addr = (uintptr_t) &av_set->coll_mc;
return FI_SUCCESS;
```

File: prov/util/src/util_coll.c [hunk 9]
```diff
@@ -1003,17 +1021,17 @@ int ofi_join_collective(struct fid_ep *ep, fi_addr_t coll_addr,
...
 
 	util_ep = container_of(ep, struct util_ep, ep_fid);
 
 	// set up the new mc for future collectives
-	new_coll_mc->mc_fid.fid.fclass = FI_CLASS_MC;
 	new_coll_mc->mc_fid.fid.context = context;
-	new_coll_mc->mc_fid.fid.ops = &util_coll_fi_ops;
-	new_coll_mc->mc_fid.fi_addr = (uintptr_t) new_coll_mc;
 	new_coll_mc->av_set = av_set;
 	new_coll_mc->ep = ep;
```
Comment (shefty) [inline prov/util/src/util_coll.c:1036]: The above 2 lines can move into the coll_mc_alloc call.  You can consider moving setting the context into there as well.

File: prov/util/src/util_coll.c [hunk 10]
```diff
@@ -1169,6 +1182,11 @@ int ofi_av_set(struct fid_av *av, struct fi_av_set_attr *attr,
 		av_set->fi_addr_count++;
 	}
 
+	ret = util_coll_mc_alloc(&av_set->coll_mc);
+	if (ret)
+		goto err2;
+	av_set->coll_mc->av_set = av_set;
```
Comment (shefty) [inline prov/util/src/util_coll.c:1188]: By embedding a coll_mc struct inside av_set, we can avoid this allocation.

File: prov/util/src/util_coll.c [hunk 11]
```diff
@@ -1169,6 +1182,11 @@ int ofi_av_set(struct fid_av *av, struct fi_av_set_attr *attr,
...
+
 	av_set->av = util_av;
 	av_set->av_set_fid.ops = &util_av_set_ops;
```
Comment (shefty) [inline prov/util/src/util_coll.c:1191]: Here is where we would want to setup the av_set->av_set_fid.fid.ops, as mentioned above.

File: prov/util/src/util_coll.c (hunk 3 above)
Comment (zhenggb72) [reply prov/util/src/util_coll.c:1029]: can you be more specific on how to change the ref count of the endpoint?

Comment (shefty) [reply prov/util/src/util_coll.c:1029]: So, yeah, we can't do that, at least not easily.  Ignore that request.

File: include/ofi_coll.h (hunk 2 above)
Comment (zhenggb72) [reply include/ofi_coll.h]: done

File: prov/util/src/util_coll.c [hunk 12]
```diff
@@ -764,6 +747,33 @@ static struct fi_ops util_coll_fi_ops = {
...
+	coll_mc->mc_fid.fid.context = context;
+	coll_mc->mc_fid.fid.ops = &util_coll_fi_ops;
+	coll_mc->mc_fid.fi_addr = (uintptr_t) coll_mc;
+	coll_mc->ep = ep;
+	assert(av_set != NULL);
+	ofi_atomic_inc32(&av_set->ref);
+	coll_mc->av_set = av_set;
+
+	return FI_SUCCESS;
+}
```
Comment (shefty) [inline prov/util/src/util_coll.c:764]: This function can just be void.

File: prov/util/src/util_coll.c [hunk 13]
```diff
@@ -1138,6 +1141,27 @@ static int util_coll_av_init(struct util_av *av)
...
 }
 
+static int util_av_set_close(struct fid *fid)
+{
+	struct util_av_set *av_set;
+
+	av_set = container_of(fid, struct util_av_set, av_set_fid.fid);
+
+	ofi_atomic_dec32(&av_set->ref);
+	assert(ofi_atomic_get32(&av_set->ref) == 0);
```
Comment (shefty) [inline prov/util/src/util_coll.c:1151]: The assert can be converted into an if statement and return -FI_EBUSY if ref > 0.  Please add a comment above the dec32 along the lines of 'release reference taken by internal coll_mc'.

File: prov/util/src/util_coll.c [hunk 14]
```diff
@@ -764,6 +747,33 @@ static struct fi_ops util_coll_fi_ops = {
...
+	coll_mc->mc_fid.fi_addr = (uintptr_t) coll_mc;
+	coll_mc->ep = ep;
+	assert(av_set != NULL);
+	ofi_atomic_inc32(&av_set->ref);
+	coll_mc->av_set = av_set;
+
+	return FI_SUCCESS;
+}
+
+int ofi_av_set_addr(struct fid_av_set *set, fi_addr_t *coll_addr)
```
Comment (shefty) [inline prov/util/src/util_coll.c:766]: This can be static.

Comment (zhenggb72) [reply prov/util/src/util_coll.c:766]: This is an API function implementation. Should it be static? I see other ofi_ functions are not static, e.g. ofi_join_collective.

Comment (shefty) [reply prov/util/src/util_coll.c:766]: If it's not used outside of this file, I would make it static.  It's exposed through a function pointer.  ofi_join_collective is accessed by a separate file.

File: prov/util/src/util_coll.c (hunk 13 above)
Comment (zhenggb72) [reply prov/util/src/util_coll.c]: I see. changed.
